### PR TITLE
#187 Iterator is a wrong concept for Fork.route()

### PR DESCRIPTION
**Pull request by dmzaytsev**
For  #187 

* created interface Opt 
* created decorator Opt.Holder 
* created Opt.Empty
* replaced Iterator<> to Opt<> in route()  for all  implementations of Fork

### CODE REVIEW

**Comment by davvd** [comment]
@dmzaytsev Many thanks, I will find someone to review it before we merge

**Comment by davvd** [comment]
@pinaf it's yours, please review

File: src/main/java/org/takes/misc/Opt.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+import lombok.EqualsAndHashCode;
+
+/**
+ * Replacement a nullable T reference with a non-null value.
+ *
+ * <p>All implementations of this interface must be immutable and thread-safe.
+ * @author Dmitry Zaytsev (dmitry.zaytsev@gmail.com)
+ * @version $Id$
+ * @since 0.13.8
```
Comment (pinaf) [inline src/main/java/org/takes/misc/Opt.java]: @dmzaytsev 0.14

File: src/main/java/org/takes/misc/Opt.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    T get();
+
+    /**
+     * Returns true if contains instance.
+     * @return True if present
+     */
+    boolean has();
+
+    /**
+     * Holder decorator.
```
Comment (pinaf) [inline src/main/java/org/takes/misc/Opt.java]: @dmzaytsev maybe say this is holder for a single element only?

**Comment by pinaf** [comment]
@dmzaytsev Holder only takes 1 element. Do you believe there is the possibility of needing more elements?

**Comment by pinaf** [comment]
@dmzaytsev good job! 2 minor comments and 1 question.

**Comment by dmzaytsev** [comment]
@pinaf I did use the interface prototype from the task #187 
```
interface Opt<T> {
  T get();
  boolean has();
}
```
I think it will be used only one element.
In a sense, this interface is an analogue  Guava Optional

**Comment by pinaf** [comment]
@dmzaytsev OK, I think I undertand it now. The user of `Fork.route` should only have no object or 1 object. That's why iterator was wrong. Is this a correct interpretation @yegor256 ?

**Comment by dmzaytsev** [comment]
@pinaf thank you for comments.
I fixed it
looks at the changes please

**Comment by pinaf** [comment]
@rultor merge

File: src/main/java/org/takes/misc/Opt.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    /**
+     * Holder for a single element only.
+     *
+     * <p>The class is immutable and thread-safe.
+     * @author Dmitry Zaytsev (dmitry.zaytsev@gamil.com)
+     * @version $Id$
+     * @since 0.14
+     */
+    @EqualsAndHashCode(of = "origin")
+    final class Holder<T> implements Opt<T> {
```
Comment (yegor256) [inline src/main/java/org/takes/misc/Opt.java]: don't like the name at all :( see http://www.yegor256.com/2015/03/09/objects-end-with-er.html every time you see -er at the end of the name, change it. How about `Single` instead?

File: src/main/java/org/takes/misc/Opt.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+     *
+     * <p>The class is immutable and thread-safe.
+     * @author Dmitry Zaytsev (dmitry.zaytsev@gamil.com)
+     * @version $Id$
+     * @since 0.14
+     */
+    final class Empty<T> implements Opt<T> {
+        @Override
+        public T get() {
+            throw new UnsupportedOperationException("#get()");
```
Comment (yegor256) [inline src/main/java/org/takes/misc/Opt.java]: I would change this `"#get()"` to something more meaningful, like `"there is nothing here, use has() first, to check"`

**Comment by yegor256** [comment]
@dmzaytsev two comments from me above

**Comment by dmzaytsev** [comment]
@yegor256 thanks for your comments
take a look the changes

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by dmzaytsev** [comment]
@yegor256 seems some problems occur with Rultor yegor256/rultor#840

**Comment by dmzaytsev** [comment]
@rultor merge again

**Comment by yegor256** [comment]
@rultor merge

**Comment by dmzaytsev** [comment]
@rultor merge

**Comment by davvd** [comment]
@elenavolokhova please, review this task for compliance with our [quality rules](http://www.teamed.io/qa.html)

**Comment by davvd** [comment]
@rultor deploy

**Comment by elenavolokhova** [comment]
@davvd This ticket looks good. Thanks, All!

**Comment by davvd** [comment]
> @davvd This ticket looks good. Thanks All!

@elenavolokhova thanks, got it, great to see high-quality tickets here

**Comment by davvd** [comment]
@pinaf 10 mins was added to the account of @elenavolokhova (for QA review), in transaction 56014233. I just added **19 mins** to your account, many thanks for your contribution.. AP-56N43257F8755032Y spent here. you're getting extra minutes here (c=4). +19 added to your rating, current score is: [+7863](http://www.netbout.com/b/35128?open=rating)

**Pull request merged**